### ISSUE TITLE
InfiniteDataLoader does not shuffle at each epoch

### ISSUE THREAD

**Issue by willyd** [ultralytics/yolov5#6741]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training

### Bug

It looks like the `InfiniteDataLoader` does not reshuffle the data when used with `num_workers!=0`. A small test program to prove my point:
```python
import torch.distributed as dist
from torch.utils.data.dataloader import DataLoader
from torch.utils.data.distributed import DistributedSampler
from utils.datasets import InfiniteDataLoader


def main(args):
    dist.init_process_group(backend="nccl" if dist.is_nccl_available() else "gloo")
    if dist.get_rank() == 0:
        print(args)

    dataset = list(range(16))
    sampler = DistributedSampler(dataset, shuffle=True)

    if args.infinite:
        loader = InfiniteDataLoader(dataset, batch_size=4, sampler=sampler, num_workers=args.num_workers, pin_memory=True)
    else:
        loader = DataLoader(dataset, batch_size=4, sampler=sampler, num_workers=args.num_workers, pin_memory=True)

    batches = {}

    for epoch in range(3):
        sampler.set_epoch(epoch)
        batches[epoch] = []
        for batch in enumerate(loader):
            batches[epoch].append(batch)

    if dist.get_rank() == 0:
        for epoch, epoch_bacthes in batches.items():
            for i, batch in enumerate(epoch_bacthes):
                print(f'rank: {dist.get_rank()}, epoch {epoch}, batch [{i:02d}] {batch}')

    dist.barrier()

    if dist.get_rank() != 0:
        for epoch, epoch_bacthes in batches.items():
            for i, batch in enumerate(epoch_bacthes):
                print(f'rank: {dist.get_rank()}, epoch {epoch}, batch [{i:02d}] {batch}')

    dist.barrier()


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--num-workers', type=int, default=4)
    parser.add_argument('--infinite', action='store_true', default=False)

    args = parser.parse_args()
    main(args)
```

Below is the output of different ways of calling the test program. If it is called with `--infinite` and `--num-workers!=0` every epoch has the same batches. Note how only the first case exhibits this behavior.

Since `InfiniteDataLoader` was introduced to avoid recreating the workers at every epoch, maybe a simple fix would be to use `DataLoader` with `persistent_workers=True`? Unless having the same order of shuffling on each epoch is what was intended?

When called like this:
```
CUDA_VISIBLE_DEVICES=2,3 python3 -m torch.distributed.run --nproc_per_node 2 test_distributed.py --num-workers=4 --infinite
```
it outputs:
```
Namespace(infinite=True, num_workers=4)
rank: 0, epoch 0, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 0, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 0, epoch 1, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 1, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 0, epoch 2, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 2, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 1, epoch 0, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 0, batch [01] (1, tensor([14,  0,  3,  1]))
rank: 1, epoch 1, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 1, batch [01] (1, tensor([14,  0,  3,  1]))
rank: 1, epoch 2, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 2, batch [01] (1, tensor([14,  0,  3,  1]))
```

When called like this:
```
CUDA_VISIBLE_DEVICES=2,3 python3 -m torch.distributed.run --nproc_per_node 2 test_distributed.py --num-workers=0 --infinite
```
it outputs:
```
Namespace(infinite=True, num_workers=0)
rank: 0, epoch 0, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 0, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 0, epoch 1, batch [00] (0, tensor([ 5,  6, 11,  7]))
rank: 0, epoch 1, batch [01] (1, tensor([ 1,  9, 10, 13]))
rank: 0, epoch 2, batch [00] (0, tensor([8, 1, 6, 0]))
rank: 0, epoch 2, batch [01] (1, tensor([ 4, 10, 15,  5]))
rank: 1, epoch 0, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 0, batch [01] (1, tensor([14,  0,  3,  1]))
rank: 1, epoch 1, batch [00] (0, tensor([15,  4,  2, 12]))
rank: 1, epoch 1, batch [01] (1, tensor([ 0,  8,  3, 14]))
rank: 1, epoch 2, batch [00] (0, tensor([13,  9,  7, 11]))
rank: 1, epoch 2, batch [01] (1, tensor([ 2, 12, 14,  3]))
```

When called like this:
```
CUDA_VISIBLE_DEVICES=2,3 python3 -m torch.distributed.run --nproc_per_node 2 test_distributed.py --num-workers=4
```
it outputs:
```
Namespace(infinite=False, num_workers=4)
rank: 0, epoch 0, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 0, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 0, epoch 1, batch [00] (0, tensor([ 5,  6, 11,  7]))
rank: 0, epoch 1, batch [01] (1, tensor([ 1,  9, 10, 13]))
rank: 0, epoch 2, batch [00] (0, tensor([8, 1, 6, 0]))
rank: 0, epoch 2, batch [01] (1, tensor([ 4, 10, 15,  5]))
rank: 1, epoch 0, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 0, batch [01] (1, tensor([14,  0,  3,  1]))
rank: 1, epoch 1, batch [00] (0, tensor([15,  4,  2, 12]))
rank: 1, epoch 1, batch [01] (1, tensor([ 0,  8,  3, 14]))
rank: 1, epoch 2, batch [00] (0, tensor([13,  9,  7, 11]))
rank: 1, epoch 2, batch [01] (1, tensor([ 2, 12, 14,  3]))
```

When called like this:
```
CUDA_VISIBLE_DEVICES=2,3 python3 -m torch.distributed.run --nproc_per_node 2 test_distributed.py --num-workers=0
```
it outputs:
```
Namespace(infinite=False, num_workers=0)
rank: 0, epoch 0, batch [00] (0, tensor([12,  9, 11, 13]))
rank: 0, epoch 0, batch [01] (1, tensor([ 2, 15,  4,  7]))
rank: 0, epoch 1, batch [00] (0, tensor([ 5,  6, 11,  7]))
rank: 0, epoch 1, batch [01] (1, tensor([ 1,  9, 10, 13]))
rank: 0, epoch 2, batch [00] (0, tensor([8, 1, 6, 0]))
rank: 0, epoch 2, batch [01] (1, tensor([ 4, 10, 15,  5]))
rank: 1, epoch 0, batch [00] (0, tensor([10,  6,  8,  5]))
rank: 1, epoch 0, batch [01] (1, tensor([14,  0,  3,  1]))
rank: 1, epoch 1, batch [00] (0, tensor([15,  4,  2, 12]))
rank: 1, epoch 1, batch [01] (1, tensor([ 0,  8,  3, 14]))
rank: 1, epoch 2, batch [00] (0, tensor([13,  9,  7, 11]))
rank: 1, epoch 2, batch [01] (1, tensor([ 2, 12, 14,  3]))
```

### Environment

- YOLO: YOLOV5 git commit 8efe97719c9b3b77c9db9b5c8592e051b7f0c9a7
- OS Ubuntu 20.04
- Python 3.8
- torch==1.10.2+cu113
- torchvision==0.11.3+cu113

### Minimal Reproducible Example

See example program above. Also attached as a TXT file
[test_distributed.py.txt](https://github.com/ultralytics/yolov5/files/8116959/test_distributed.py.txt)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@willyd thanks for the bug report! This would be a serious problem if you are right. I'll try to reproduce shuffle errors on COCO128 today.

**Comment by glenn-jocher** [comment]
@willyd if I delete all but 4 images from COCO128 and then run `train.py --noval --epochs 3` I get this when printing `index`:
https://github.com/ultralytics/yolov5/blob/601dbb83f01b58355211f2565cfa4eecb48b1220/utils/datasets.py#L554-L555

```bash
     Epoch   gpu_mem       box       obj       cls    labels  img_size
  0%|          | 0/1 [00:00<?, ?it/s]
0
3
1
2
       0/2        0G   0.04684    0.0348   0.08958        17       320: 100%|██████████| 1/1 [00:03<00:00,  3.86s/it]

     Epoch   gpu_mem       box       obj       cls    labels  img_size
  0%|          | 0/1 [00:00<?, ?it/s]
1
3
2
0
       1/2        0G   0.05717   0.03431   0.06143        18       320: 100%|██████████| 1/1 [00:00<00:00,  1.02it/s]

     Epoch   gpu_mem       box       obj       cls    labels  img_size
  0%|          | 0/1 [00:00<?, ?it/s]
2
3
1
0
       2/2        0G   0.04935   0.03459    0.1005        19       320: 100%|██████████| 1/1 [00:00<00:00,  1.01it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95:   0%|          | 0/1 [00:00<?, ?it/s]
0
1
2
3
```

**Comment by glenn-jocher** [comment]
@willyd if I set all augmentation hyps to zero and capture the first 5 train_batch.jpg images (which will correspond to epochs 0-4, they look like this. The first two batches are identical, (EDIT: 2 and 4 also) which is pretty strange, but the rest are random. These are with mosaic switched off also. With mosaic these images would be the top left image in the mosaic and the other 3 would be randomly drawn from the entire set.

0
![train_batch0](https://user-images.githubusercontent.com/26833433/156748572-f0ada313-9316-4230-9945-cca089026668.jpg)

1
![train_batch1](https://user-images.githubusercontent.com/26833433/156748636-17f67aac-5865-4263-b71a-a220730df687.jpg)

2
![train_batch2](https://user-images.githubusercontent.com/26833433/156748652-461f08e2-c072-4fef-a6cf-afa508e6f88d.jpg)

3
![train_batch3](https://user-images.githubusercontent.com/26833433/156748670-0db55e82-0349-4c65-a25c-206326dfa3b9.jpg)

4
![train_batch4](https://user-images.githubusercontent.com/26833433/156748683-31f9bdf2-d464-4d6b-9504-adc14b7fb68b.jpg)

**Comment by glenn-jocher** [comment]
@willyd repeated above experiment with 9 image dataset. I'm not seeing any repetition at all now.

0
![train_batch0](https://user-images.githubusercontent.com/26833433/156749304-947d7a02-9721-472b-b0a5-da375e3bbc3e.jpg)

1
![train_batch1](https://user-images.githubusercontent.com/26833433/156749318-4098b4d0-d50a-49d8-8685-43f688fbe10f.jpg)

2
![train_batch2](https://user-images.githubusercontent.com/26833433/156749323-e3486b85-8d68-43fa-ace4-634fecbecd2b.jpg)

3
![train_batch3](https://user-images.githubusercontent.com/26833433/156749346-ad9bf3a2-8ee3-43a1-98f9-0deef3290998.jpg)

4
![train_batch4](https://user-images.githubusercontent.com/26833433/156749359-aed8f247-f3dc-4e48-8edd-90d52b5b274c.jpg)

**Comment by glenn-jocher** [comment]
@willyd repeated with 18-image dataset at batch-size 9, also saw no repetitions. My conclusion is I'm not able to independently reproduce any shuffle issues with the trainloader.

Regardless, if you can simplify the code with your proposed fix then it still merits consideration. 

> Since InfiniteDataLoader was introduced to avoid recreating the workers at every epoch, maybe a simple fix would be to use DataLoader with persistent_workers=True? Unless having the same order of shuffling on each epoch is what was intended?

The fastest and easiest way to incorporate your ideas into the official codebase is to submit a Pull Request (PR) implementing your idea, and if applicable providing before and after profiling/inference/training results to help us understand the improvement your feature provides. This allows us to directly see the changes in the code and to understand how they affect workflows and performance.

Please see our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/) to get started.

**Comment by willyd** [comment]
@glenn-jocher Thanks for checking this issue. I will revisit my observations based on your comments and get back to you. Thanks.

**Comment by willyd** [comment]
@glenn-jocher I updated my script to support a single GPU and single process mode and there is no repetition in that case, i.e. shuffling is different at every epoch. I think the issue only happens when the `DistributedSampler` is used. 

From your comment:

> @willyd if I delete all but 4 images from COCO128 and then run train.py --noval --epochs 3 I get this when printing index:

I understand that you only checked proper shuffling in single GPU mode? May I suggest you try with more than one GPU and see what you get?

**Comment by glenn-jocher** [comment]
@willyd yes I only checked single-GPU. Ok I will check DDP training, thanks!

**Comment by glenn-jocher** [comment]
@willyd I've been waiting to test on our Multi-GPU machine but it's occupied with trainings at the moment. 

Can you please submit a PR with your proposed fix to allow shuffling for DDP trainings? Thanks! It would be interesting to see if this improves results on DDP trainings.

**Comment by willyd** [comment]
@glenn-jocher I still need to verify that the conclusions I drew from a simplified example hold for the complete DDP YOLOv5 training loop. I suppose I could replicate the small tests you did to see. If I observe the same behavior, then I could submit a PR.

How much testing do you expect from me for a PR like this? The introduction of the InfiniteDataLoader seems to have required a lot of training time. Like you I can't dedicate a lot of GPU time to this kind of work.

**Comment by glenn-jocher** [comment]
@willyd I've created a PR myself in #6920. Is this the main change you had in mind?

I think now that we have a PR it will be easier to test, i.e. 

```
git clone https://github.com/ultralytics/yolov5 -b update/dataloader
```

EDIT: will require a lot of testing. I can do this once some GPUs free up.

**Comment by willyd** [comment]
@glenn-jocher It's precisely what I had in mind. Thanks for putting this together.

**Comment by willyd** [comment]
@glenn-jocher I just checked and there does not seem to be any repetition in the batches in DDP when executing this command:
```
 CUDA_VISIBLE_DEVICES=2,3 python -m torch.distributed.launch --nproc_per_node 2 train.py --img 640 --batch-size 8 --epochs 4 --data coco128.yaml --weights yolov5s.pt --cache --noval > log_master.txt
 ```
Keeping only 8 images in coco128 and printing index like you did I get (for rank == 0):
```
Rank: 0, Epoch: 0
Rank: 0, index: 4
Rank: 0, index: 7
Rank: 0, index: 2
Rank: 0, index: 1
Rank: 0, Epoch: 1
Rank: 0, index: 5
Rank: 0, index: 2
Rank: 0, index: 7
Rank: 0, index: 1
Rank: 0, Epoch: 2
Rank: 0, index: 0
Rank: 0, index: 7
Rank: 0, index: 6
Rank: 0, index: 1
Rank: 0, Epoch: 3
Rank: 0, index: 2
Rank: 0, index: 3
Rank: 0, index: 1
Rank: 0, index: 6
```
and for rank == 1
```
Rank: 1, Epoch: 0
Rank: 1, index: 0
Rank: 1, index: 3
Rank: 1, index: 5
Rank: 1, index: 6
Rank: 1, Epoch: 1
Rank: 1, index: 4
Rank: 1, index: 6
Rank: 1, index: 3
Rank: 1, index: 0
Rank: 1, Epoch: 2
Rank: 1, index: 4
Rank: 1, index: 2
Rank: 1, index: 5
Rank: 1, index: 3
Rank: 1, Epoch: 3
Rank: 1, index: 4
Rank: 1, index: 5
Rank: 1, index: 0
Rank: 1, index: 7
```
I suppose there must be a bug in my small test script that led me to think that the indices would repeat themselves in the YOLOv5 training loop too. I think we can close this issue. Sorry for the noise.
